distribution/lib/Standard/Database/0.0.0-dev/src/Extensions/Upload_In_Memory_Table.enso

```diff
@@ -120,6 +120,67 @@ Table.select_into_database_table self connection (table_name : Text) primary_key
      More expensive checks, like clashing keys or unmatched rows are checked
      only on a sample of rows, so errors may still occur when the output action
      is enabled.
-Table.update_database_table : Database_Table -> Update_Action -> Vector Text | Nothing -> Boolean -> Problem_Behavior -> Database_Table ! Table_Not_Found | Unmatched_Columns | Missing_Input_Columns | Column_Type_Mismatch | SQL_Error | Illegal_Argument
-Table.update_database_table self target_table (update_action : Update_Action = Update_Action.Update_Or_Insert) (key_columns : Vector | Nothing = default_key_columns target_table) (error_on_missing_columns : Boolean = False) (on_problems : Problem_Behavior = Problem_Behavior.Report_Warning) =
-    common_update_table self target_table update_action key_columns error_on_missing_columns on_problems
+Table.update_rows : Table | Database_Table -> Update_Action -> Vector Text | Nothing -> Boolean -> Problem_Behavior -> Database_Table ! Table_Not_Found | Unmatched_Columns | Missing_Input_Columns | Column_Type_Mismatch | SQL_Error | Illegal_Argument
+Table.update_rows self (source_table : Table | Database_Table) (update_action : Update_Action = Update_Action.Update_Or_Insert) (key_columns : Vector | Nothing = Nothing) (error_on_missing_columns : Boolean = False) (on_problems : Problem_Behavior = Problem_Behavior.Report_Warning) =
+    _ = [source_table, update_action, key_columns, error_on_missing_columns, on_problems]
+    Error.throw (Illegal_Argument.Error "Table.update_rows modifies the underlying table, so it is only supported for Database tables - in-memory tables are immutable.")
```

Comment:
I'm not sure this is a good idea. I appreciate a potential usefulness of such operation, but the practical API is vastly different.

In the Database, what we do is we mutably modify the target table (and return it back). This is a side-effecting operation. From now on, all references to the table will have updated data.

In in-memory, Tables are immutable, so all this would do is return a _new_ table with modified contents. A bit like DB dry-run (but currently our dry-run returns the _original_ table **unchanged** - so even here this does not match).

Reply:
We aim for our APIs to be compatible and mostly interchangeable between DB and in-memory, of course to the extent possible.

I'm afraid that introducing an operation which behaves in a similar but fundamentally different and subtle way is asking for trouble - when switching between in-memory and Database the users may not notice the subtle differences and get bad results.

 -----

I'm not completely sure. To be honest, the difference is indeed rather small and it feels like in practice in many workflows it will not be felt. So it seems like this could work.

But I think it can introduce hard to understand errors (with the unexpected behaviour change).

IMHO it is better to have a _clear_ distinction between side-effecting _WRITE_ operations (like this one, or writing to file etc. that update some _external state), and normal 'immutable' transformations creating a new modified data structure (which this would have to become for the in-memory approach to work).

If we really want the 'search and replace' behaviour, maybe we should introduce a separate method, that returns a _new modified_ table with these changes applied? We could then implement it both for in-memory and Database and ensure that in both cases they do not modify the original table, but just return a new one (i.e. in Database it would be a modified query which does the replace, essentially a kind of a VIEW; I guess it may not be most efficient in that backend, but at least would be consistent).

What do you think? IMO creating a _single_ operation that for one backend is a _side-effecting WRITE_ and for another is an _immutable TRANSFORM_ is risky. If anything, I'd simply create a separate operation for the TRANSFORM scenario.

Reply:
It's a bit extreme, but we could append _write to every method that writes to the database, so it would always be clear.